clients/common/hipblas_gentest.py

```diff
@@ -0,0 +1,574 @@
+
+    def mark_str(mark):
+        line = source[mark.line]
+        return("In file " + line[1] + ", line " + str(line[2]) + ", column " +
+               str(mark.column + 1) + ":\n" + line[0].rstrip() + "\n" +
+               ' ' * mark.column + "^\n")
+
+    # We iterate through all of the documents to properly diagnose errors,
+    # because the load_all generator does not handle exceptions correctly.
+    docs = []
+    load = Loader(source_str)
+    while load.check_data():
+        try:
+            doc = load.get_data()
+        except yaml.YAMLError as err:
+            sys.exit((mark_str(err.problem_mark) if err.problem_mark else "") +
+                     (err.problem + "\n" if err.problem else "") +
+                     (err.note + "\n" if err.note else ""))
+        else:
+            docs.append(doc)
+    return docs
+
+
+def get_datatypes(doc):
+    """ Get datatypes from YAML doc"""
+    dt = ctypes.__dict__.copy()
+    for declaration in doc.get('Datatypes') or ():
+        for name, decl in declaration.items():
+            if isinstance(decl, dict):
+                # Create derived class type based on bases and attr entries
+                dt[name] = type(name,
+                                tuple([eval(t, dt)
+                                       for t in decl.get('bases') or ()
+                                       if TYPE_RE.match(t)]
+                                      ), decl.get('attr') or {})
+                # Import class' attributes into the datatype namespace
+                for subtype in decl.get('attr') or {}:
+                    if TYPE_RE.match(subtype):
+                        dt[subtype] = eval(name+'.'+subtype, dt)
+            elif isinstance(decl, str) and TYPE_RE.match(decl):
+                dt[name] = dt[decl]
+            else:
+                sys.exit("Unrecognized data type "+name+": "+repr(decl))
+    return dt
+
+
+def get_arguments(doc):
+    """The kernel argument list, with argument names and types"""
+    return [(var, eval(decl[var], datatypes))
+            for decl in doc.get('Arguments') or ()
+            if len(decl) == 1
+            for var in decl
+            if TYPE_RE.match(decl[var])]
+
+
+def setkey_product(test, key, vals):
+    """Helper for setdefaults. Tests that all values in vals is present
+    in test, if so then sets test[key] to product of all test[vals]."""
+    if all(x in test for x in vals):
+        result = 1
+        for x in vals:
+            if x in ('incx', 'incy'):
+                result *= abs(test[x])
+            else:
+                result *= test[x]
+        test[key] = int(result)
+
+
+def setdefaults(test):
+    """Set default values for parameters"""
+    # Do not put constant defaults here -- use hipblas_common.yaml for that.
+    # These are only for dynamic defaults
+    # TODO: This should be ideally moved to YAML file, with eval'd expressions.
+
+    if test['function'] in ('asum_strided_batched', 'nrm2_strided_batched',
```

Comment:
It's a little strange that we're using underscores in function names when the functions are actually CamelCase. This is prevalent in other places in the hipBLAS clients, though, so maybe a future task to change them all.

Reply:
It is example code it won't be true unless we add test or bench yaml with to those names I to think.
We might consider no smart defaults at all.  This just adds the system, actual use will be another PR.  I thought maybe hipblas_smoke.yaml might be first use case?